### PR TITLE
JWT authentication with signature and payload validation

### PR DESCRIPTION
**Pull request by zvonand**
Continues #66058.  

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added support for JWT authentication of existing users

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: src/Access/ExternalAuthenticators.cpp [hunk 1]
```diff
@@ -547,6 +649,25 @@ HTTPAuthClientParams ExternalAuthenticators::getHTTPAuthenticationParams(const S
...
 }
 
+bool ExternalAuthenticators::checkJWTCredentials(const String &claims, const JWTCredentials & credentials, SettingsChanges &settings) const
+{
+    std::lock_guard lock{mutex};
+
+    const auto token = String(credentials.getToken());
+    const auto &user_name = credentials.getUserName();
+
+    for (const auto &it : jwt_validators)
```
Comment (ilejn) [inline src/Access/ExternalAuthenticators.cpp:659]: CH codestyle suggests spaces around '&' here

**Comment by alexey-milovidov** [comment]
@zvonand a test for LDAP has failed: https://s3.amazonaws.com/clickhouse-test-reports/70660/75916285535e3855b62ed481661613e3eef2fbe0/integration_tests__aarch64__[5_6].html

**Comment by zvonand** [comment]
@alexey-milovidov yes, I know. There is an issue for this which was reopened https://github.com/ClickHouse/ClickHouse/issues/65762

I just don't have more ideas of why it is failing yet, hope to come up with something

**Comment by ilejn** [comment]
Hello @zvonand , have a look at wait_ldap_to_start .
Not sure if it checks anything besides container is running, because it does not pay attention at
result code of self.exec_in_container . 
Am I missing something?

**Comment by zvonand** [comment]
@thevar1able you were assigned on the predecessor PR: #66058

Maybe you could also take a look at this one?

**Comment by melvynator** [comment]
Thank you for taking the time to submit this pull request. We appreciate your interest in contributing to our project.

After review this PR have a few issues:

1. There were some implementation issues raised by our security team
2. It doesn't match all of our requirements that we have for this feature (needed for our cloud)

Because of the two points above it conflicts with our internal implementations and prevent us from being able to merge it. 

You can see that this is a feature we are working on internally in the roadmap https://github.com/ClickHouse/ClickHouse/issues/58392 (labeled with a cloud logo).

**Comment by zvonand** [comment]
@melvynator @alexey-milovidov does this mean you are going to open-source it? Or it will be forever cloud-only?

**Comment by melvynator** [comment]
@zvonand We don't have plan to open-source it for now.

**Comment by aadant** [comment]
@melvynator this is a feature that is also relevant to non-cloud users. will it be the same with OIDC / OAuth2 https://github.com/ClickHouse/ClickHouse/issues/28383  ?

**Comment by zvonand** [comment]
> There were some implementation issues raised by our security team

@melvynator it would be nice if you could outline these concerns or their ideas (as it may not only be relevant to this exact PR, but to some other stuff or some future development)

**Pull request closed without merging**